### PR TITLE
Removing Layer groups removed when no layers left in them

### PR DESCRIPTION
**Pull request by SeqLaz**
Removing Layer groups removed when no layers left in them

Test project:

[crop_monitoring_000.zip](https://github.com/user-attachments/files/17775665/crop_monitoring_000.zip)

Implemented on QFieldSync plugin on the PR https://github.com/opengisch/qfieldsync/pull/623

Fix https://github.com/opengisch/qfieldsync/issues/587

### CODE REVIEW

File: libqfieldsync/offline_converter.py [hunk 1]
```diff
@@ -346,6 +350,18 @@ def _convert(self, project: QgsProject) -> None:
         QgsProject.instance().write(str(export_project_filename))
         project.writeProject.disconnect(on_original_project_write)
 
+    def remove_empty_groups_from_project(self, group) -> None:
```
Comment (suricactus) [inline libqfieldsync/offline_converter.py:353]:
```suggestion
    def remove_empty_groups_from_project(self, group: Qgs...) -> None:
```

Please add typings.

File: libqfieldsync/offline_converter.py [hunk 2]
```diff
@@ -346,6 +350,18 @@ def _convert(self, project: QgsProject) -> None:
...
+        """
+        Recursively removes any empty groups from the given layer tree group.
+        """
+        for child in group.children():
+            if isinstance(child, QgsLayerTreeGroup):
+                self.remove_empty_groups_from_project(child)
+
+                if not child.children():
+                    print(f"Removing empty group: {child.name()}")
+                    group.removeChildNode(child)
```
Comment (suricactus) [inline libqfieldsync/offline_converter.py:363]:
```suggestion
        for child in group.children():
            if not isinstance(child, QgsLayerTreeGroup):
                continue

            self.remove_empty_groups_from_project(child)

            if not child.children():
                print(f"Removing empty group: {child.name()}")
                group.removeChildNode(child)
```

This pattern is called "early return", or in our case, "early continue". It makes the code less nested and shows clear intention what kind of context you want to manage. You can find it across QFieldSync a lot, so whenever possible and makes sense - try to use it :)

File: libqfieldsync/offline_converter.py (hunk 1 above)
Comment (suricactus) [reply libqfieldsync/offline_converter.py:353]:
```suggestion
    def remove_empty_groups_from_project(self, project: QgsProject) -> None:
```

File: libqfieldsync/offline_converter.py [hunk 3]
```diff
@@ -279,6 +280,9 @@ def _convert(self, project: QgsProject) -> None:
             elif layer_action == SyncAction.REMOVE:
                 project.removeMapLayer(layer)
 
+        project_layers_groups = project.layerTreeRoot()
+        self.remove_empty_groups_from_project(project_layers_groups)
```
Comment (suricactus) [inline libqfieldsync/offline_converter.py:284]:
```suggestion
        self.remove_empty_groups_from_project(project)
```

**Review by suricactus — CHANGES_REQUESTED** [review]
Good work! A few small changes and it is ready to go!

Don't forget to open QFieldSync PR with the new `libqfieldsync` commit sha.

Also make a reference to the fixed issue on QFieldSync here and on the `QFieldSync` PR.

And don't forget to make cross reference between the two PRs.

File: libqfieldsync/offline_converter.py (hunk 1 above)
Comment (SeqLaz) [reply libqfieldsync/offline_converter.py:353]: Unfortunately, this approach didn't work for nested groups.

To handle it recursively, it's better to pass the group directly. If you have a better idea of how to, please let me know.

File: libqfieldsync/offline_converter.py (hunk 2 above)
Comment (SeqLaz) [reply libqfieldsync/offline_converter.py:363]: Great, indeed is much better!

File: libqfieldsync/offline_converter.py (hunk 3 above)
Comment (SeqLaz) [reply libqfieldsync/offline_converter.py:284]: Completely right!

File: libqfieldsync/offline_converter.py [hunk 4]
```diff
@@ -346,6 +349,18 @@ def _convert(self, project: QgsProject) -> None:
...
 
+    def remove_empty_tree_groups_from_project(self, group: QgsLayerTreeGroup) -> None:
+        """
+        Recursively removes any empty groups from the given layer tree group.
+        """
+        for child in group.children():
+            if not isinstance(child, QgsLayerTreeGroup):
+                continue
+
+            self.remove_empty_tree_groups_from_project(child)
```
Comment (suricactus) [inline libqfieldsync/offline_converter.py:360]:
```suggestion

	    # remove recursively
            self.remove_empty_groups_from_layer_tree_group(child)

```

Sorry, I overlooked this recursion here. Place make it explicit and add a line before and after just to highlight it in the function body.

Also a new suggestion for the name of the function, sorry for being annoying with renames.

Feel free to force push with all in a single commit.

Comment (SeqLaz) [reply libqfieldsync/offline_converter.py:360]: Great!... Thanks for the suggestions!

**Comment by SeqLaz** [comment]
Hey @suricactus I addressed the suggestions!.... Also for tests I added the project sample in the descriptions of the PR

File: libqfieldsync/offline_converter.py (hunk 1 above)
Comment (suricactus) [reply libqfieldsync/offline_converter.py]: Great, I missed the recursion. Note to myself to read more carefully!

File: libqfieldsync/offline_converter.py [hunk 5]
```diff
@@ -346,6 +349,21 @@ def _convert(self, project: QgsProject) -> None:
...
+        """
+        Recursively removes any empty groups from the given layer tree group.
+        """
+        for child in group.children():
+            if not isinstance(child, QgsLayerTreeGroup):
+                continue
+
+            # remove recursively
+            self.remove_empty_groups_from_layer_tree_group(child)
+            if not child.children():
```
Comment (suricactus) [inline libqfieldsync/offline_converter.py:364]:
```suggestion
            self.remove_empty_groups_from_layer_tree_group(child)

            if not child.children():
```

**Review by suricactus — APPROVED** [review]
One last empty line :)

Please squash into a single commit and we are ready to go!

**Comment by SeqLaz** [comment]
> One last empty line :)
> 
> Please squash into a single commit and we are ready to go!

Great!... Apologize my bad! :)

**Comment by SeqLaz** [comment]
Hey @suricactus I think now is ready to go!

**Review by suricactus — APPROVED** [review]
Great! Don't forget to open a PR on QFieldSync with the new `libqfieldsync` version, cross reference this PR to the one on QFieldSync and vice versa, and put in both "fixes <LINK_TO_GITHUB_ISSUE>". Ideally this should be done when you open the PR, not just before closing :) .

**Pull request merged**